Allow to create local RRef with value

aazzolini opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28948 Allow to create local RRef with value**

Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

aazzolini committed:
Allow to create local RRef with value

Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -154,6 +170,9 @@ RRefForkData RRefContext::prepareChildFork(const std::shared_ptr<RRef>& rref) {
     addForkOfOwner(rfd.rrefId_, rfd.forkId_);
+    // ensure that this RRef is in the owners_ list to keep it alive.
+    // this is needed for OwnerRRefs that were created locally.
+    owners_[rref->rrefId()] = rref;
```
It seems we need a lock here to guard `owners_`.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -2,7 +2,7 @@
 from torch.distributed import _start_rpc_agent
 from torch.distributed import _destroy_rref_context, _cleanup_python_rpc_handler
-from torch.distributed import WorkerInfo
+from torch.distributed import WorkerInfo, RRef as PyRRef
```
Why do we need `as PyRRef` here?

aazzolini replied:
I wanted to expose RRef to rpc.RRef  -- but avoid the lint issues for RRef not being used in the file. I can alternatively add a lint ignore rule. Is this how you'd do it?

mrshenli replied:
Yep, either adding a `# noqa: F401` or do something similar to this PR #25823

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -67,6 +67,7 @@ PyObject* rpc_init(PyObject* /* unused */) {
               "local_value",
               &PyRRef::localValue,
               py::call_guard<py::gil_scoped_release>())
+          .def(py::init(&PyRRef::local))
```
Move this to top as this is a constructor?

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -140,6 +140,22 @@ template std::shared_ptr<OwnerRRef<IValue>> RRefContext::getOrCreateOwnerRRef<
     getOrCreateOwnerRRef<py::object>(const RRefId& rrefId);
 
+template <typename T>
+std::shared_ptr<OwnerRRef<T>> RRefContext::createOwnerRRef() {
```
I assume that this does not take a `T` argument as we can also use this to support run `rpc.remote` on self?

aazzolini replied:
I agree that we can use this to run rpc.remote on self, but I don't follow how it would be possible to remove the T argument -- I need to pass it to the constructor of OnwerRRef.

mrshenli replied:
Oh, I mean I was thinking this API should take a `T&& value` for now as this is how we use it currently, i.e., something like:

```c++
template <typename T>
std::shared_ptr<OwnerRRef<T>> RRefContext::createOwnerRRef(T&& value) {
```

aazzolini replied:
Yes i'm thinking of a use case where we don't set the value right away.

aazzolini committed:
Update on "Allow to create local RRef with value"


Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

[ghstack-poisoned]

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -914,6 +922,50 @@ def test_rref_forward_chain(self):
 
+    @dist_init
+    def test_local_rref_no_fork(self):
+        local_rref = rpc.RRef(35)
```
This API is a little weird. This RRef has nothing to do with rpc, yet we're creating it with `rpc.RRef()`. I think RRef should be a separate package since it is independent from RPC (as we can see here). If we want to allow creating local RRefs like this, RRef shouldn't be tightly coupled with RPC.

aazzolini replied:
There's 2 things in your comment:

1. would you prefer to have simply torch.distributed.RRef ?  I can do that.

2. Currently init_rpc() needs to be called before we can call RRef() constructor even locally since only OnwerRRef constructor is private and only accessible by RRefContext. I can change that too if you think is appropriate? In that case, RRefs would become known to RRefContext only at fork time.

pritamdamania87 replied:
1. Yes, I think that makes sense. 
2. I think this is why we probably need a higher level initialization like `init_model_parallel` (I'm open to a better name) that initializes rpc, rref and autograd. The user has to call this API even if they want to only use rpc/rref/dist_autograd individually.

aazzolini replied:
I thought a bit about this -- I don't think that it makes sense to initialize RRef without initializing RPC, actually. RRef semantic requires RPC concepts such as owner() . owner() only makes sense if an RPCAgent is present.

RPC without RRef doesn't make sense either -- since RRef is a native type supported as both input and output of any RPC calls, particularly Remote.

As for 1., I'll change it.

pietern replied:
Please don't do 1. The top level `torch.distributed` package concerns only collectives. As you say, RPC without RRef and RRef without RPC are both useless. If that ever changes, we could consider promoting RRefs to their own package, but not to the top level `torch.distributed` module.

pietern replied:
Update: I see they're part of `torch.distributed` today, but that's because that's where they were defined from the very beginning. I have #27529 out (pending type checking fixes) to update that and move the pybind11 wrappers to the appropriate modules. We can discuss moving them around, but let's use a new issue for that.

aazzolini replied:
Ok i'm just going to reuse the existing distributed.RRef for now. Will leave to @pietern to move it afterwards.

pritamdamania87 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -67,6 +67,7 @@ PyObject* rpc_init(PyObject* /* unused */) {
               "local_value",
               &PyRRef::localValue,
```
Just realized that these are all part of the public user API and we should have good documentation for all of these. We should define all of these methods in api.py and have proper python docs. Don't have to do this as part of this PR.

pritamdamania87 commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -152,6 +152,13 @@ PyRRef PyRRef::unpickle(const py::tuple& t) {
   return PyRRef(std::move(rref));
 }
 
+PyRRef PyRRef::local(const py::object& value) {
```
We do we have a static method like this instead of having a constructor?

aazzolini replied:
The constructor of PyRRef is not exposed to python. This is something that works on top of that constructor.

pritamdamania87 replied:
Can't we define a constructor for PyRRef that takes a py::object and use that in py::init()?

aazzolini committed:
Update on "Allow to create local RRef with value"


Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

[ghstack-poisoned]

aazzolini committed:
Update on "Allow to create local RRef with value"


Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

[ghstack-poisoned]

aazzolini committed:
Update on "Allow to create local RRef with value"


Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

[ghstack-poisoned]

pietern commented:
@aazzolini Please hold off on merging until #27529 has landed. I've been trying to merge it for a while and it should be any moment now...

pietern commented:
@aazzolini The PR has been merged. If you rebase you can use `rpc.RRef`. See #27529 and b4df413712098ff93da694472d95896e62faf07d.

aazzolini committed:
Update on "Allow to create local RRef with value"


Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

[ghstack-poisoned]

mrshenli commented on test/rpc_test.py:
```diff
@@ -916,6 +925,50 @@ def test_rref_forward_chain(self):
+            rpc.remote(
+                dst_worker,
+                add_rref_to_value,
+                args=(rref, 50)).to_here().wait(), 90)
```
This is currently flaky. Can you explicitly hold the rref alive for now?

mrshenli replied:
I am landing #29396, which will revert `to_here()` and make it block again.

mrshenli reviewed (approved):
Looks good from my side! Thanks!

aazzolini committed:
Update on "Allow to create local RRef with value"


Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

[ghstack-poisoned]

aazzolini committed:
Update on "Allow to create local RRef with value"


Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

[ghstack-poisoned]
